CIP-XXXX/README.md

```diff
@@ -0,0 +1,173 @@
+
+Additionally, apps need to be able to submit a transaction to the wallet for signing in an asynchronous manner, as gathering of signatures can take a long time and each wallet provider will have its own way of handling this process.
+
+Finally, the signTx() and signData() endpoints will have to be disabled when using this extension since they are not compatible with native script based addresses.
+
+## Rationale for the required data
+- Script descriptor: Any transaction consuming a UTxO from a native script based address must attach the corresponding script.
+- Script Requirements list: 
+	-- dApps need to know the number of signers to calculate the fees correctly. 
+	-- Including the correct "required signers" brings UX improvements. 
+	-- dApps need to know if the script has any before-after requirements to set the correct validBefore and TTL values.
+- Collateral donator Native script based addresses cannot provide collateral for Plutus transactions.
+
+### Additional API Endpoints
+
+#### api.getCollateralAddress(): Promise\<Address>
+
+For Plutus V2 and later, partial collateral is supported. This function returns an address that can be used to add collateral to a transaction. The address returned must be owned by one of the signers in the list of signers returned by `api.getScriptRequirements()`. 
+
+dApp developers can choose to use this address to add collateral to a transaction, or they can choose to use the `api.getCollateral()` function to get a list of UTXOs that can be used as collateral. If the dApp chooses to use this address, they must ensure that the address is not used for any other purpose, as the wallet may be using it to track collateral, and that the collateral return address is the same one.
+
+#### api.getScriptRequirements: Promise\<ScriptRequirement[]>
+
+Errors: `APIError`
+
+Returns a list of ScriptRequirements that will be used to validate any transaction sent to the wallet.
+
+#### api.getScript(): Promise\<cbor\<nativeScript>>
+
+Errors: `APIError`
+
+Returns the CBOR-encoded native script that controls this wallet. 
+
+#### api.submitUnsignedTx(tx: cbor\<unsignedTransaction>): Promise\<hash32>
+
+Errors: `APIError`, `TxError`
+
+Submits a transaction to the wallet for signing. The wallet should check that the transaction is valid, gather the required signatures, compose the finalized transaction, and submit the transaction to the network. If the transaction is valid and the wallet is able to sign it, the wallet should return the transaction hash. If the transaction is invalid or the wallet is unable to sign it, the wallet should throw a `TxError` with the appropriate error code. The wallet should not submit the transaction to the network if it is invalid or the wallet is unable to sign it.
+
+If the transaction contains hidden metadata, the wallet should not submit the transaction when it is ready, but return it to the dApp when the dApp calls the `getCompletedTx` function.
```

Comment:
I don't understand this point, could you please elaborate? Is it about witness datum? If so, why can't the dapp provide a partial witness set with it's own witnesses when calling this method?

Reply:
Metadata not datum. 

when using sighTx() you are able to attach the hash of the transaction metadata, sighing the hash makes the transaction valid, but to submit the transaction you need the actual metadata

Reply:
Ah I see, thanks. In that case, why can't the dapp provide auxiliary data when signing?

Reply:
I could be the intension to hide the metadata from the user.

I created a multisig NFTs minting  service for example, but you don't want the user to know the details of the NFT they are about to mint until the Tx is on-chain.

Reply:
Sounds shady but I'll trust it's a legit use case :joy:

If we can find a way to keep sign and submit separate, then it's a non-issue and would make it simpler.

Reply:
>If we can find a way to keep sign and submit separate, then it's a non-issue and would make it simpler. 

This endpoint is where we separate them  🤓

But for regular operation having it all in one go, is much better UX and DevX.

Reply:
> This endpoint is where we separate them 🤓

Hm, I understand `If the transaction contains hidden metadata, the wallet should not submit the transaction when it is ready` that otherwise wallet can/should submit it after it's signed.

Reply:
> But for regular operation having it all in one go, is much better UX and DevX.

I can see the appeal though. Maybe it could be an explicit argument/option set by the dapp instead, like `{ submit: boolean }`. What I don't like in the current design is that it's likely this logic would have to be extended, in the future (like applying similar logic for datums/redeemers if dapp wants to keep them secret).

Reply:
Good point, I was not aware this is even possible with datums and redeemers.